### ISSUE TITLE
Incorrect missing property error when not using strict null checks

### ISSUE THREAD

**Issue by paulvanbrenk** [Microsoft/TypeScript#29862]
[PlayGround link](https://www.typescriptlang.org/play/#src=type%20ItDone%20%3D%20%7B%20done%3A%20true%20%7D%3B%0D%0Atype%20ItValue%20%3D%20%7B%20value%3A%20any%3B%20done%3A%20false%20%7D%3B%0D%0Atype%20ItResult%20%3D%20ItDone%20%7C%20ItValue%3B%0D%0A%0D%0Alet%20x%20%3D%2012%3B%0D%0Afunction%20next()%3A%20ItResult%20%7B%0D%0A%20%20%20%20return%20x%20%3D%3D%3D%2012%20%3F%20%7B%20done%3A%20true%20%7D%20%3A%20%7B%20value%3A%20x%2C%20done%3A%20false%20%7D%3B%0D%0A%7D%0D%0A%0D%0Alet%20y%20%3D%20next()%3B%0D%0Aif%20(!y.done)%20%7B%0D%0A%20%20%20%20alert(y.value)%3B%0D%0A%7D)

``` typescript
type ItDone = { done: true };
type ItValue = { value: any; done: false };
type ItResult = ItDone | ItValue;

let x = 12;
function next(): ItResult {
    return x === 12 ? { done: true } : { value: x, done: false };
}

let y = next();
if (!y.done) {
    alert(y.value); // with strict null checks enabled this compiles, without 'value' is reported as missing on ItDone
}
```

I expected this to work the same with and without strict null checks.

**Comment by jack-williams** [comment]
Falseyness is not a discriminant of `true | false` without strict null checks because both `undefined` and `null` inhabit `true`.

```ts
function next(): ItResult {
    // compiles fine without --strictNullChecks
    return x === 12 ? { done: undefined } : { value: x, done: false };
}
```

**Comment by paulvanbrenk** [comment]
I assumed (wrongly) that by typing done with either `true` or `false` that wouldn't be allowed. 

Thanks!

(Closing)
